service/listeners.go

```diff
@@ -0,0 +1,235 @@
+// Copyright 2024 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package service
+
+import (
+	"errors"
+	"io"
+	"net"
+	"sync"
+	"sync/atomic"
+
+	"github.com/Jigsaw-Code/outline-sdk/transport"
+)
+
+// The implementations of listeners for different network types are not
+// interchangeable. The type of listener depends on the network type.
+type Listener = io.Closer
+
+// StreamListener is a network listener for stream-oriented protocols that
+// accepts [transport.StreamConn] connections.
+type StreamListener interface {
+	// Accept waits for and returns the next connection to the listener.
+	AcceptStream() (transport.StreamConn, error)
+
+	// Close closes the listener.
+	// Any blocked Accept operations will be unblocked and return errors.
+	Close() error
+
+	// Addr returns the listener's network address.
+	Addr() net.Addr
+}
+
+type acceptResponse struct {
+	conn transport.StreamConn
+	err  error
+}
+
+type sharedListener struct {
+	listener    net.TCPListener
+	acceptCh    chan acceptResponse
+	closeCh     chan struct{}
+	onCloseFunc func() error
+}
+
+// Accept accepts connections until Close() is called.
+func (sl *sharedListener) AcceptStream() (transport.StreamConn, error) {
+	select {
+	case acceptResponse := <-sl.acceptCh:
+		if acceptResponse.err != nil {
+			return nil, acceptResponse.err
+		}
+		return acceptResponse.conn, nil
```

Comment:
You need to check for channel closed too, as a way to communicate the underlying listener is closed:

```suggestion
	case acceptResponse, ok := <-sl.acceptCh:
		if !ok {
			return nil, net.ErrClosed
		}
		if acceptResponse.err != nil {
			return nil, acceptResponse.err
		}
		return acceptResponse.conn, nil
```